firewood/src/merkle/changes.rs

```diff
@@ -0,0 +1,389 @@
+        if len == 0 {
+            return None;
+        }
+        // Generate a random index within the bounds of the set's size
+        let index = rng.random_range(0..len);
+        // Use the iterator's nth method to get the element at that index
+        set.iter().nth(index).cloned()
+    }
+
+    fn gen_delete_key(
+        rng: &SeededRng,
+        committed_keys: &HashSet<Vec<u8>>,
+        seen_keys: &mut HashSet<Vec<u8>>,
+    ) -> Option<Vec<u8>> {
+        let key_opt = random_key_from_hashset(rng, committed_keys);
+        // Just skip the key if it has been seen. Otherwise return it and add the key
+        // to seen keys.
+        key_opt
+            .filter(|del_key| !seen_keys.contains(del_key))
+            .inspect(|del_key| {
+                seen_keys.insert(del_key.clone());
+            })
+    }
+
+    fn gen_random_keys(
+        rng: &SeededRng,
+        committed_keys: &HashSet<Vec<u8>>,
+        num_keys: usize,
+        start_val: usize,
+    ) -> (Vec<BatchOp<Vec<u8>, Box<[u8]>>>, usize) {
+        const CHANCE_DELETE: usize = 2;
+        let mut seen_keys = std::collections::HashSet::new();
+        let mut i = 0;
+        let mut batch = Vec::new();
+
+        while batch.len() < num_keys {
+            if !committed_keys.is_empty() && rng.random_range(1..=100) <= CHANCE_DELETE {
+                let del_key = gen_delete_key(rng, committed_keys, &mut seen_keys);
+                if let Some(key) = del_key {
+                    batch.push(BatchOp::Delete { key });
+                    continue;
+                }
+                // If we couldn't generate a delete key, then just fall through and create
+                // a BatchOp::Put.
+            }
+            let key_len = rng.random_range(1..=32);
+            let key: Vec<u8> = (0..key_len).map(|_| rng.random()).collect();
+            // Only add if key is unique
+            if seen_keys.insert(key.clone()) {
+                batch.push(BatchOp::Put {
+                    key,
+                    value: Box::from(format!("value{}", i + start_val).as_bytes()),
+                });
+            }
+            i += 1;
+        }
+        (batch, i + start_val)
+    }
+
+    #[test]
+    fn test_preorder_iterator() {
+        let rng = firewood_storage::SeededRng::from_env_or_random();
+        let (batch, _) = gen_random_keys(&rng, &HashSet::new(), 1000, 0);
+
+        // Keep a sorted copy of the batch.
+        let mut batch_sorted = batch.clone();
+        batch_sorted.sort_by_key(|op| op.key().clone());
+
+        // Insert batch into a merkle trie.
+        let mut merkle = create_test_merkle();
+        for item in &batch {
+            // All of the ops should be Puts
+            merkle
+                .insert(
+                    item.key(),
+                    item.value().unwrap().to_vec().into_boxed_slice(),
+                )
+                .unwrap();
+        }
+        let merkle = make_immutable(merkle);
```

Comment:
```suggestion
        // freeze and compute the hash
        let merkle: Merkle<NodeStore<Arc<ImmutableProposal>, MemStore>> = merkle.into().unwrap();
```
this function is only called from here, so a comment and doing this inline is probably easier. I had to check what make_immutable does, and it's just a simple into() method call with an unwrap.

Historical note: this used to be called `freeze` before it was a fallable type conversion.